### DISCUSSION TITLE
About convolutional layers : `nn.Conv2d`

### DISCUSSION THREAD

**Discussion by HyeAnn-Lee** [N1-314/Intrinsic-Curiosity-Module---PyTorch#9] [General]
PyTorch 1.12.1 버전의 nn.Conv2d 클래스의 [docs](https://pytorch.org/docs/1.12/generated/torch.nn.Conv2d.html?highlight=conv2d#torch.nn.Conv2d)는 input과 output의 shape 이렇게 말하고 있습니다.

![image](https://github.com/user-attachments/assets/ebfd2e9f-a175-4bf8-87ca-bf45180063ff)

[ICM 논문](https://arxiv.org/pdf/1705.05363)을 보면, 아래와 같이 언급이 되어 있습니다.

> 3. Experimental Setup
>> Training details
>>> The input RGB images are converted into gray-scale and re-sized to 42 × 42. In order to model temporal dependencies, the state representation ($s_t$) of the environment is constructed by concatenating the current frame with the three previous frames.

초기 input $(N, C_{in}, H_{in}, W_{in})$이 $(4, 1, 42, 42)$ 라는 뜻입니다.

> 3. Experimental Setup
>> A3C architecture
>>> The input state $s_t$ is passed through a sequence of _four convolution layers with 32 filters each, kernel size of 3 × 3, stride of 2 and padding of 1_. An exponential linear unit (ELU) is used after each convolution layer.
>> Intrinsic Curiosity Module (ICM) architecture
>>> The inverse model first maps the input state ($s_t$) into a feature vector $\phi(s_t)$ using a series of _four convolution layers, each with 32 filters, kernel size 3 × 3, stride of 2 and padding of 1_. ELU non-linearity is used after each convolution layer. The dimensionality of $\phi(s_t)$ (i.e. the output of the fourth convolution layer) is 288.

32 filter, kernel 3, stride 2, padding 1인 convolutional layer를 42x42에 손으로 직접 적용하려고 하면 한두칸씩 오차가 나기 때문에 논문의 설명을 의심할 수도 있습니다. 하지만 PyTorch는 위 이미지의 floor 함수를 통해 이를 잘 처리하고 있는 것으로 보입니다.

논문의 값들을 적용하면, $H_{out} = floor ((H_{in}+1)/2)$ 가 되어 이미지의 크기가 42->21->11->6->3으로 줄어듦을 알 수 있습니다.

**Comment by HyeAnn-Lee** [comment]
는 아니었구연..,,,,,,,,
